File: probfilter-core/src/main/scala/probfilter/pdsa/cuckoo/CuckooTable.scala
package probfilter.pdsa.cuckoo

import probfilter.util.JavaFriendly

import scala.reflect.ClassTag


/**
 * An immutable cuckoo hash table for cuckoo filters.
 *
 * @see [[probfilter.pdsa.cuckoo.TypedCuckooTable]]
 */
trait CuckooTable extends Serializable {
  def typed[T]: TypedCuckooTable[T] = this.asInstanceOf[TypedCuckooTable[T]]
}


object CuckooTable {
  def empty[T: ClassTag]: CuckooTable = TypedCuckooTable.empty[T]

  @JavaFriendly(scalaDelegate = "probfilter.pdsa.cuckoo.CuckooTable.empty")
  def emptyByte: CuckooTable = empty[Byte]

  @JavaFriendly(scalaDelegate = "probfilter.pdsa.cuckoo.CuckooTable.empty")
  def emptyShort: CuckooTable = empty[Short]

  @JavaFriendly(scalaDelegate = "probfilter.pdsa.cuckoo.CuckooTable.empty")
  def emptyLong: CuckooTable = empty[Long]
}

File: probfilter-core/src/main/java/com/c6h5no2/probfilter/pdsa/cuckoo/CuckooTable.java
package com.c6h5no2.probfilter.pdsa.cuckoo;

import java.io.Serializable;


/**
 * A cuckoo hash table for cuckoo filters.
 *
 * @see TypedCuckooTable
 */
public sealed interface CuckooTable extends Serializable permits TypedCuckooTable {
    @SuppressWarnings("unchecked")
    default <T> TypedCuckooTable<T> typed() {
        return (TypedCuckooTable<T>) this;
    }
}
